### ISSUE TITLE
Using same input and output port causes request duplication

### ISSUE THREAD

**Issue by walczakp** [buger/goreplay#292]
Hello! Amazing work you did here. However, we found a small issue at our company.

Using gor with this commandline: 
`gor --output-http-timeout '1800s' --input-raw 'intf-ip-addr:8080' --output-http 'http://output-http-addr:8080|100%' --http-allow-url '/some/url'`
causes requests to be replayed multiple times on target machine (as checked on tomcat access-logs and tcpdump).
Replaying request to another port (eg. 8081) fixes the issue. We confirmed the issue on gor versions 0.12.1 and 0.13.0, however, issue is not present in version 0.10.1.
Machine runs binary release of gor on CentOS 6.6 2.6.32-504.12.2.el6.x86_64.

**Comment by buger** [comment]
Thank you for the report! Did you see any warning errors from Gor output?
When you tell multiple times, do you see any correlation like always 2 times or 3 times?

Indeed weird issue.

**Comment by pewusoft** [comment]
On 0.12.1 - there were many "broken pipe" errors (still, in tomcat access logs multiple identical request could be seen). On 0.13.0 however, the verbose output from Gor was nearly silent. Only --output-http-stats was notifying about the problem (numbers like 40-50 req/sec, when we expected something around 10).

Unfortunately, we noticed that some requests were replayed correctly, while another could be replayed even 8 times too much (with that value varying from 2).

If you couldn't reproduce the issue, I'll try to do this on totally different environment, outside of my working hours. Maybe it's only environmental.

**Comment by buger** [comment]
Thanks! Are those requests GET, POST or maybe 301 redirects? Or it happens for all kinds of traffic?

**Comment by pewusoft** [comment]
Most definitely GETs, there are no redirects in our workflow, however we have currently no way to extinguish POST requests from each other. We'll try to perform a more specific test when it would be possible.

**Comment by pewusoft** [comment]
What is more, I managed to reproduce the issue using three CentOS 6.6 VM's (kernel 2.6.32-642.1.1.el6.x86_64). With server set up using `python -m SimpleHTTPServer 8080` on two machines, I set up gor to replicate requests from one to another. When I fired just one request using curl to original machine, machine set as gor output was flooded with endless loop of requests.

**Comment by buger** [comment]
Thanks! I made some changes, can you check this binaries:
[gor_ip-check_mac.tar.gz](https://github.com/buger/gor/files/303084/gor_ip-check_mac.tar.gz)
[gor_ip-check_x64.tar.gz](https://github.com/buger/gor/files/303085/gor_ip-check_x64.tar.gz)

And branch itself https://github.com/buger/gor/tree/listener-ip-check

**Comment by pewusoft** [comment]
Unfortunately, now it didn't replay any request at all. However, I noticed a different problem. When I sent a request (`curl 10.0.2.6:8080`) from machine which was running gor (everything as in my previous example, but request originator was 10.0.2.5) to gor output machine (10.0.2.6 as in my example), gor caught this request and also entered an infinite loop of requests. This happenend on both 0.13 release version, and your custom binaries from previous post.
I don't know Go, but from what I see, you added some check not to replay packets with target IP. I don't know using libpcap either, but could the issue be some parameter causing libpcap to listen packets with not only source port set to desired value (eg. 8080), but also target port? This could make libpcap "hear" packets that gor replayed, thus making infinite loop. Again, this is only a blind shot, as I know nothing about Go or libpcap.

**Comment by buger** [comment]
I stably replicated issue, working on fixing. 👍

**Comment by buger** [comment]
I think I found the issue, here is the binary:

[gor_DEV-1465402913_x64.tar.gz](https://github.com/buger/gor/files/305216/gor_DEV-1465402913_x64.tar.gz)

**Comment by pewusoft** [comment]
Great work! I'm happy to confirm that both of issues I found are gone using this binary.

**Comment by wk8** [comment]
Hit that issue too, and happy to confirm your bin above fixed it :) would you please mind making it into a release?

**Comment by buger** [comment]
@wk8 definitely! Should happen in next few days.

**Comment by wk8** [comment]
Thanks! and thanks again for gor, it's a great piece of software :)

**Comment by buger** [comment]
Released in https://github.com/buger/gor/releases/tag/v0.14.0

**Comment by wk8** [comment]
Wow, that was fast, thanks a lot!

**Comment by hieuvuvan** [comment]
on latest version 1.2.0, i got this
tried 1.1.0 did not have

**Comment by buger** [comment]
Will re-open! Let us confirm if it is a regression.

**Comment by buger** [comment]
@hieuvuvan can you post the exact command of how you run GoReplay?

**Comment by hieuvuvan** [comment]
i tested it using docker-compose.
- my purpose capture request port 80 nginx replay to echoserver port 80 - just httpserver receive request, send info response and log it. 
- i download gor version 1.2.0 on nginx container and use command `./gor --input-raw :80 --output-http echoserver:80` to replay
- when i make only one request from a container alpine on same docker-compose network to nginx but i see echoserver endless loop
- i try explicit ip at option `--input-raw` but got same result 
- tried version 1.1.0 not get this
```yaml
# docker-compose.yml
version: '2'
services:
  echoserver:
    image: hieuvv/echoserver
    environment:
    - PORT=80
    - DEBUG=true
    dns:
    - 8.8.8.8
  nginx:
    image: nginx:alpine
    dns:
    - 8.8.8.8
  alpine:
    image: alpine
    command: sleep infinity
    dns:
    - 8.8.8.8
```
result test:
![result test](https://user-images.githubusercontent.com/15248282/94632120-fecba700-02f3-11eb-808c-db4b0b81a6a3.png)

**Comment by urbanishimwe** [comment]
@hieuvuvan AFAIK, all containers in compose join a single network with a single internal IP. if that is not the case, it's worth showing the internal configuration of your compose. otherwise, this will be closed as explained in #828

**Comment by hieuvuvan** [comment]
i dont think i get problem same as #828. AFAIK, docker-compose allocate difference ip for containers
when
> i try explicit ip at option --input-raw

i used nginx container ip to replay to echoserver container ip with same port 80. i also test this case at version 1.1.0 and did not get this problem

this is my result of `docker network inspect`
```
[
    {
        "Name": "test_default",
        "Id": "298eb67b231c5449d2efe298da38ca74fa3a56fdb13bcdc56eb6e3df51381fc0",
        "Created": "2020-09-30T18:19:51.8429643+07:00",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.17.0.0/16",
                    "Gateway": "172.17.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": true,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {
            "0b6a6467d107c5510351fc2dad222264eccd0bdc61a91964baa9a6e8387f515c": {
                "Name": "test_nginx_1",
                "EndpointID": "07f91ad96122cf5062f916d579ca2aa88421a61ad2f791c8aca360a445c21b51",
                "MacAddress": "02:42:ac:11:00:03",
                "IPv4Address": "172.17.0.3/16",
                "IPv6Address": ""
            },
            "d520f6f2c342c2cae1f8b6aeb334a28ade8b4fe78fe2ed5e63fc59a11d38c69e": {
                "Name": "test_echoserver_1",
                "EndpointID": "c12981b6a36e7310085811e38ff04798fa09d67b092dc0741bad8d56421859f7",
                "MacAddress": "02:42:ac:11:00:02",
                "IPv4Address": "172.17.0.2/16",
                "IPv6Address": ""
            },
            "f941383e728ef47e4931ee0541969409eb75b342249de80a04dc0307db12ebd2": {
                "Name": "test_alpine_1",
                "EndpointID": "30a858e27d1509d499d665a9df96209f7bed050b4b74493d060b7004a4f2f49e",
                "MacAddress": "02:42:ac:11:00:04",
                "IPv4Address": "172.17.0.4/16",
                "IPv6Address": ""
            }
        },
        "Options": {},
        "Labels": {
            "com.docker.compose.network": "default",
            "com.docker.compose.project": "test",
            "com.docker.compose.version": "1.27.3"
        }
    }
]
````

**Comment by mmfrb** [comment]
Hi there, in case you want more examples, I'm also having this problem in v1.2.0, but it doesn't happen when downgrading to v1.1.0.

I tested everything in a Kubernetes Cluster.

Dockerfile:
```
FROM alpine:latest
RUN wget https://github.com/buger/goreplay/releases/download/v1.1.0/gor_1.1.0_x64.tar.gz
RUN tar -xzf gor_1.1.0_x64.tar.gz
``` 

Pod with echoserver + goreplay that redirects to internal service.
```
apiVersion: v1
kind: Pod
metadata:
  name: goreplay-test
spec:
  containers:
    - name: echoserver
      image: ealen/echo-server:latest
      ports:
        - containerPort: 80
      env:
        - name: PORT
          value: "80"
    - name: goreplay
      image: GOREPLAY_DOCKER_IMAGE
      command: ["./gor"]
      args: ["--input-raw", ":80", "--output-http", "INTERNAL_SERVICE"]
      securityContext:
        capabilities:
          add: ["CAP_NET_ADMIN"]
```

Pod with another echoserver that receives mirrored requests:
```
apiVersion: v1
kind: Pod
metadata:
  name: goreplay-test-mirror
  labels:
    run: goreplay-test-mirror
spec:
  containers:
    - name: echoserver
      image: ealen/echo-server:latest
      ports:
        - containerPort: 80
      env:
        - name: PORT
          value: "80"
```

Hope this helps. I'm using 1.1.0 for now.

**Comment by mooba** [comment]
Hi, I also got problem in v1.2.0, I tested with tcpdump and confirm it did replay twice when receiving one request. I'll try to use 1.1.0 refer the discussion above

**Comment by hmih** [comment]
I'm hitting this as well on 1.3, 1.2, and as suggested using 1.1 fixed it

**Comment by urbanishimwe** [comment]
@hmih with the command you've used, the predictable/fair outcome is the one you're seeing in 1.3>!
if you know the specific `IP` you want to listen on you may try `-input-raw ${IP-ADDRESS}:80`

**Comment by buger** [comment]
Should be fixed with fresh 1.3 release.
